// bs-platform/lib/es6/js_option.js


import * as Js_primitive from "./js_primitive.js";

function some(x) {
  return Js_primitive.some(x);
}

function isSome(param) {
  return param !== undefined;
}

function isSomeValue(eq, v, x) {
  if (x !== undefined) {
    return eq(v, Js_primitive.valFromOption(x));
  } else {
    return false;
  }
}

function isNone(param) {
  return param === undefined;
}

function getExn(x) {
  if (x !== undefined) {
    return Js_primitive.valFromOption(x);
  } else {
    throw new Error("getExn");
  }
}

function equal(eq, a, b) {
  if (a !== undefined) {
    if (b !== undefined) {
      return eq(Js_primitive.valFromOption(a), Js_primitive.valFromOption(b));
    } else {
      return false;
    }
  } else {
    return b === undefined;
  }
}

function andThen(f, x) {
  if (x !== undefined) {
    return f(Js_primitive.valFromOption(x));
  }
  
}

function map(f, x) {
  if (x !== undefined) {
    return Js_primitive.some(f(Js_primitive.valFromOption(x)));
  }
  
}

function getWithDefault(a, x) {
  if (x !== undefined) {
    return Js_primitive.valFromOption(x);
  } else {
    return a;
  }
}

function filter(f, x) {
  if (x !== undefined) {
    var x$1 = Js_primitive.valFromOption(x);
    if (f(x$1)) {
      return Js_primitive.some(x$1);
    } else {
      return undefined;
    }
  }
  
}

function firstSome(a, b) {
  if (a !== undefined) {
    return a;
  } else if (b !== undefined) {
    return b;
  } else {
    return undefined;
  }
}

var $$default = getWithDefault;

export {
  some ,
  isSome ,
  isSomeValue ,
  isNone ,
  getExn ,
  equal ,
  andThen ,
  map ,
  getWithDefault ,
  $$default ,
  $$default as default,
  filter ,
  firstSome ,
  
}
/* No side effect */
